dbms/src/Storages/Distributed/DistributedBlockOutputStream.cpp

```diff
@@ -80,6 +80,40 @@ void DistributedBlockOutputStream::writePrefix()
 
 
 void DistributedBlockOutputStream::write(const Block & block)
+{
+    auto has_materialized = [&]()
+    {
+        for (const auto & col : storage.getColumns().getMaterialized())
+            if (block.has(col.name))
+                return true;
+        return false;
+    };
+
+    if (has_materialized())
+        writeSyncAsync(eraseMaterializedFromBlock(block));
+    else
+        writeSyncAsync(block);
+}
+
+/* They are added by the AddingDefaultBlockOutputStream, and we will get
+ * different number of columns eventually */
+Block DistributedBlockOutputStream::eraseMaterializedFromBlock(const Block & block) const
+{
+    Block res{ block };
+
+    for (const auto & col : storage.getColumns().getMaterialized())
+        if (res.has(col.name))
+        {
+            res.erase(col.name);
+            LOG_DEBUG(log, storage.getTableName()
+                << ": column " + col.name + " will be removed, "
+                << "because it is MATERIALIZED");
+        }
+
+    return res;
+}
+
+void DistributedBlockOutputStream::writeSyncAsync(const Block & block)
```

Comment:
The name is hilarious :)
Let's just add the loop with `block.erase(...)` to `write()`, it's short enough so we don't have to split it.

Reply:
>The name is hilarious :)

Indeed)

> Let's just add the loop with block.erase(...) to write()

On this case `Block` will be always copied

Reply:
copying `Block` is almost free, the patch has been simplified